sdk/secp256r1/src/lib.rs

```diff
@@ -0,0 +1,615 @@
+//! Instructions for the [secp256r1 native program][np].
+//!
+//! [np]: https://docs.solana.com/developing/runtime-facilities/programs#secp256r1-program
+solana_pubkey::declare_id!("Secp256r1SigVerify1111111111111111111111111");
+
+use bytemuck::{Pod, Zeroable};
+
+#[derive(Default, Debug, Copy, Clone, Zeroable, Pod, Eq, PartialEq)]
+#[repr(C)]
+pub struct Secp256r1SignatureOffsets {
+    /// Offset to compact secp256r1 signature of 64 bytes
+    pub signature_offset: u16,
+
+    /// Instruction index where the signature can be found
+    pub signature_instruction_index: u16,
+
+    /// Offset to compressed public key of 33 bytes
+    pub public_key_offset: u16,
+
+    /// Instruction index where the public key can be found
+    pub public_key_instruction_index: u16,
+
+    /// Offset to the start of message data
+    pub message_data_offset: u16,
+
+    /// Size of message data in bytes
+    pub message_data_size: u16,
+
+    /// Instruction index where the message data can be found
+    pub message_instruction_index: u16,
+}
+
+#[cfg(all(not(target_arch = "wasm32"), not(target_os = "solana")))]
+mod target_arch {
+    use {
+        crate::Secp256r1SignatureOffsets,
+        bytemuck::bytes_of,
+        openssl::{
+            bn::{BigNum, BigNumContext},
+            ec::{EcGroup, EcKey, EcPoint},
+            ecdsa::EcdsaSig,
+            nid::Nid,
+            pkey::PKey,
+            sign::{Signer, Verifier},
+        },
+        solana_feature_set::FeatureSet,
+        solana_instruction::Instruction,
+        solana_precompile_error::PrecompileError,
+    };
+
+    pub const COMPRESSED_PUBKEY_SERIALIZED_SIZE: usize = 33;
+    pub const SIGNATURE_SERIALIZED_SIZE: usize = 64;
+    pub const SIGNATURE_OFFSETS_SERIALIZED_SIZE: usize = 14;
+    pub const SIGNATURE_OFFSETS_START: usize = 2;
+    pub const DATA_START: usize = SIGNATURE_OFFSETS_SERIALIZED_SIZE + SIGNATURE_OFFSETS_START;
+
+    // Order as defined in SEC2: 2.7.2 Recommended Parameters secp256r1
+    pub const SECP256R1_ORDER: [u8; 32] = [
+        0xFF, 0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xBC, 0xE6, 0xFA, 0xAD, 0xA7, 0x17, 0x9E, 0x84, 0xF3, 0xB9, 0xCA, 0xC2, 0xFC, 0x63,
+        0x25, 0x51,
+    ];
+
+    // Computed SECP256R1_ORDER - 1
+    pub const SECP256R1_ORDER_MINUS_ONE: [u8; 32] = [
+        0xFF, 0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xBC, 0xE6, 0xFA, 0xAD, 0xA7, 0x17, 0x9E, 0x84, 0xF3, 0xB9, 0xCA, 0xC2, 0xFC, 0x63,
+        0x25, 0x50,
+    ];
+
+    // Computed half order
+    const SECP256R1_HALF_ORDER: [u8; 32] = [
+        0x7F, 0xFF, 0xFF, 0xFF, 0x80, 0x00, 0x00, 0x00, 0x7F, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xDE, 0x73, 0x7D, 0x56, 0xD3, 0x8B, 0xCF, 0x42, 0x79, 0xDC, 0xE5, 0x61, 0x7E, 0x31,
+        0x92, 0xA8,
+    ];
+
+    pub fn new_secp256r1_instruction(
+        message: &[u8],
```

Comment:
Sorry I missed this from yesterday, but was there a reason we generate the signing key internally in the function (and throwing it away at the end) instead of taking it in as input? For this function to be useful, it seems like it would need to either take in a key from the outside or return the key as part of the output.

Reply:
Absolutely correct. I got too absorbed in making it work for the testing use cases to remember its meant for external use as well. Fixing 👍